kernel/src/sched.rs

```diff
@@ -881,20 +883,99 @@ impl Kernel {
+                // process accessible memory. If a process passes an
+                // invalid pointer (outside of its memory regions),
+                // the MPU must catch this memory access and fault the
+                // process accordingly.
+                let fn_ptr: Option<NonNull<()>> = NonNull::new(upcall_ptr);
+
+                // A upcall is identified as a tuple of the driver
+                // number and the subdriver number.
+                let upcall_id = UpcallId {
+                    driver_num: driver_number,
+                    subscribe_num: subdriver_number,
+                };
+
+                // Construct the upcall struct
+                let upcall = Upcall::new(process.processid(), upcall_id, appdata, fn_ptr);
+
+                let rval = platform.with_driver(driver_number as usize, |driver| match driver {
+                    Some(d) => {
+                        let res =
+                            d.subscribe(subdriver_number as usize, upcall, process.processid());
+
+                        match res {
+                            Err((returned_upcall, err)) => {
+                                // The capsule has refused the subscribe operation,
+                                // verify that it passed back the new upcall
+                                if returned_upcall.app_id != process.processid()
+                                    || returned_upcall.upcall_id != upcall_id
+                                    || returned_upcall.appdata != appdata
+                                    || returned_upcall.fn_ptr != fn_ptr
+                                {
+                                    // The capsule did not return the Upcall passed in
+                                    //
+                                    // TODO: How to handle this?
+                                    panic!(
+                                        "Driver {}, subscribe num {}: Upcall swapped in error case",
+                                        driver_number, subdriver_number
+                                    );
+                                } else {
+                                    // Capsule returned the correct Upcall
+
+                                    // TODO: The capsule might have already scheduled
+                                    // upcalls on the new instance, and we must not
+                                    // clear upcalls on the previous instance in
+                                    // this branch (capsule refused the subscribe
+                                    // operation). However, if the two upcalls are
+                                    // identical, we can't distinguish the two and
+                                    // will cancel upcalls on _both_ the previous
+                                    // and the new instance.
+                                    process.remove_pending_upcalls(returned_upcall.upcall_id);
```

Comment:
I will take a stance on this and say that I think it is fine if we clear upcalls on the previous instance in this branch in the case where the rejected upcall was identical to an already held upcall. If a process calls `subscribe()` for some `(driver_num, subscribe_num)` and the capsule rejects, I think it is reasonable for the app to expect no future upcalls associated with that tuple. Can we just edit the subscribe TRD to state as such?

Reply:
I'm still somewhat convinced we can work around that, without too many hacks and good ergonomics. There are some TODOs in there still, for instance currently a capsule can reject subscribe returning the old upcall, and accept subscribe returning the new upcall -- both of these should be prohibited. By some clever type wrapping the passed upcall and enforcing these invariants by wrapping the `swap` method, we can effectively prevent all of this wrong behavior. Can we give this a shot first, and not immediately declare this as "works as intended"? We should be able to quickly decide on either approach once this is merged.

Reply:
Ok, I am fine to leave this as a TODO. But is it a blocking item for 2.0?

Reply:
It's supposedly a quick and small change, and I think the implications of upcalls being cleared if subscribe is accepted, but the with the new upcall returned can lead to inconsistencies in user space, if not unsoundness. So I think it's important that we enforce that the subscribe return values are actually consistent.